runtime/controller.go

```diff
@@ -118,6 +126,165 @@ func NewController(ctx context.Context, rt *Runtime, instanceID string, logger *
 	return c, nil
 }
 
+type broadcastHandler struct {
+	zapHandler    *zapslog.Handler
+	clients       map[chan *runtimev1.Log]struct{}
+	instanceID    string
+	mu            sync.Mutex
+	messages      []*runtimev1.Log
+	bufferSize    int64
+	maxBufferSize int64
+}
+
+func (b *broadcastHandler) Enabled(ctx context.Context, level slog.Level) bool {
+	return b.zapHandler.Enabled(ctx, level)
+}
+
+func (b *broadcastHandler) Handle(ctx context.Context, record slog.Record) error {
+	record.AddAttrs(slog.String("instance_id", b.instanceID))
+	size := 0
+	attrs := make(map[string]any)
+	gatherAttr := func(attr slog.Attr) bool {
+		attrs[attr.Key] = attr.Value.Any()
+		size += len(attr.Key) + len(attr.Value.String()) // approx size
+		return true
+	}
+	// hacky way of collecting attributes
+	record.Attrs(gatherAttr)
+	payload, err := structpb.NewStruct(attrs)
+	if err != nil {
+		return err
+	}
+	size += len(record.Message)
+	// add enum size, assuming upper bound for 64 bits system
+	size += 8
+	// add proto Timestamp size
+	size += 12
+	go func() {
+		message := &runtimev1.Log{
+			Level:   slogLevelToPBLevel(record.Level),
+			Time:    timestamppb.New(record.Time),
+			Message: record.Message,
+			Payload: payload,
+			Size:    uint32(size),
+		}
+		b.broadcast(message)
+	}()
+	return b.zapHandler.Handle(ctx, record)
+}
+
+func slogLevelToPBLevel(level slog.Level) runtimev1.LogLevel {
+	switch level {
+	case -4:
+		return runtimev1.LogLevel_LOG_LEVEL_DEBUG
+	case 0:
+		return runtimev1.LogLevel_LOG_LEVEL_INFO
+	case 4:
+		return runtimev1.LogLevel_LOG_LEVEL_WARN
+	case 8:
+		return runtimev1.LogLevel_LOG_LEVEL_ERROR
+	default:
+		return runtimev1.LogLevel_LOG_LEVEL_UNSPECIFIED
+	}
+}
+
+func (b *broadcastHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	return b.zapHandler.WithAttrs(attrs)
+}
+
+func (b *broadcastHandler) WithGroup(name string) slog.Handler {
+	return b.zapHandler.WithGroup(name)
+}
+
+func (b *broadcastHandler) addClient(client chan *runtimev1.Log, replay bool) {
+	b.mu.Lock()
+	b.clients[client] = struct{}{}
+	messages := append([]*runtimev1.Log(nil), b.messages...)
+	b.mu.Unlock()
+
+	if replay {
+		// Send historical messages in a non-blocking way.
+		go func() {
+			for _, msg := range messages {
+				client <- msg
+			}
+		}()
+	}
+}
+
+func (b *broadcastHandler) removeClient(client chan *runtimev1.Log) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	delete(b.clients, client)
+	close(client) // close the channel here after the client is removed.
+}
+
+func (b *broadcastHandler) broadcast(message *runtimev1.Log) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.bufferSize += int64(message.Size)
+	b.messages = append(b.messages, message)
+
+	// Trim the old messages if buffer size exceeds the limit.
+	for b.bufferSize > b.maxBufferSize && len(b.messages) > 0 {
+		removedMsg := b.messages[0]
+		b.messages = b.messages[1:]
+		b.bufferSize -= int64(removedMsg.Size)
+	}
+
+	for client := range b.clients {
+		client <- message
+	}
+}
```

Comment:
I'm wondering if there's a way to avoid continuously re-allocating the messages buffer? Some ideas are a) also having a limit on the number of messages and using a circular buffer, or b) serializing logs to a bytes format and having one circular bytes buffer of length `maxBufferSize` (that would also require changing the RPCs).

What do you think?

Reply:
a. how much limit would be put, we will have to come up with some upper bound which will depend on the size of logs buffer size b. If we serialize and send the serilized stream, wouldn't client need to handle that ? What about dynamic linked list not backed by a slice so memory is reclaimed when log is truncated but it will have memory overhead of pointers.